src/engine/engine/Entity.hpp

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <glm/glm.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+#include <glm/gtc/type_ptr.hpp>
+
+#include "engine/Component.hpp"
+#include "engine/Model.hpp"
+#include "engine/Shader.hpp"
+
+#include <unordered_map>
+
+namespace lei3d
+{
+    struct Transform
+    {
+        glm::vec3 position;
+        glm::vec3 rotation;       
+        glm::vec3 scale;
+    };
+
+    class Component;
+    class Model;
+    class Shader;
+
+    class Entity
+    {
+    private:
+        std::unordered_map<std::string, std::vector<Component>> m_ComponentLookup;   //typename, Component
+        std::vector<Component> m_Components;
+    public:
+        Model* m_Model = nullptr;    
+        Transform transform;
+        Shader* m_Shader = nullptr;
+
+        Entity();
+        Entity(Model* model);
+        Entity(Model* model, Shader* shader);
+        ~Entity();
+
+        void Start();
+        void Update(float deltaTime);
+        void PhysicsUpdate(float deltaTime);
+        void Render();
+        void OnDestroy();
+
+        void SetPosition(glm::vec3 translation);
+        void SetScale(glm::vec3 scale);
+
+        void SetShader(Shader* shader);
+
+        //Template component stuff
+
+        template<typename C>
+        C GetComponent() {
+            static_assert(std::is_same<C, Component>::value, "C must be a component type");
```

Comment:
Definitely in the updated version, add some doc comments that explain that the data is being stored internally by the function.

Reply:
i.e. don't call the constructor directly so AddComponent can create type and use the init functions to essentially do the constructor behavior.